static_src/components/env_panel.jsx

```diff
@@ -0,0 +1,219 @@
+      .then(() => envActions.fetchEnv(app.guid));
+  }
+
+  handleShowUserEnvClick(e) {
+    e.preventDefault();
+    this.setState(({ showUserEnv, showAddForm }) => ({
+      showUserEnv: !showUserEnv,
+      showAddForm: !showUserEnv ? false : showAddForm
+    }));
+  }
+
+  handleAddFormDismiss(e) {
+    e.preventDefault();
+    this.setState(() => ({ showAddForm: false }));
+    const { app } = this.props;
+    envActions.invalidateUpdateError(app.guid);
+  }
+
+  handleShowEnvClick(e) {
+    e.preventDefault();
+    this.setState(({ showEnv }) => ({ showEnv: !showEnv }));
+  }
+
+  renderUserVarItems() {
+    const { app, env, updateError } = this.props;
+    const { environment_json: envVars } = env;
+
+    return Object.keys(envVars).map(name => (
+      <EnvVarListItem
+        key={`${name}${envVars[name]}`}
+        name={name}
+        value={envVars[name]}
+        app={app}
+        updateError={updateError}
+        deleteError={updateError}
+        onUpdate={this.handleUpdate}
+        onDelete={this.handleDelete}
+      />
+    ));
+  }
+
+  renderAddAction() {
+    const { app } = this.props;
+    const { updating } = app;
+
+    const label = 'Add env var';
+    return (
+      <Action
+        clickHandler={this.handleAddClick}
+        label={label}
+        type="outline"
+        disabled={updating}
+      >
+        {label}
+      </Action>
+    );
+  }
+
+  renderShowHideAction(show, handler) {
+    return (
+      <Action
+        style="primary"
+        clickHandler={handler}
+        label={show ? 'Hide' : 'Reveal'}
+        type="outline"
+      >
+        {show ? 'Hide' : 'Reveal'}
+      </Action>
+    );
+  }
+
+  render() {
+    const { styler } = this;
+    const { app, env, updateError } = this.props;
+    const { showUserEnv, showAddForm, showEnv } = this.state;
+    const { updating } = app;
+
+    return (
+      <div>
+        <Section styler={styler}>
+          <Header styler={styler}>User-defined environment variables</Header>
+          <PanelActions>
+            {this.renderShowHideAction(
+              showUserEnv,
+              this.handleShowUserEnvClick
+            )}
+            {showUserEnv && !showAddForm && this.renderAddAction()}
+          </PanelActions>
+        </Section>
+        {showUserEnv && (
+          <div>
+            {showAddForm && (
+              <EnvVarForm
+                disabled={updating}
+                onDismiss={this.handleAddFormDismiss}
+                onSubmit={this.handleUpdate}
+              />
+            )}
+            {showAddForm && updateError && <ErrorMessage error={updateError} />}
+            <ComplexList>{this.renderUserVarItems()}</ComplexList>
+          </div>
+        )}
+        <Section styler={styler}>
+          <Header styler={styler}>System-defined environment variables</Header>
+          {this.renderShowHideAction(showEnv, this.handleShowEnvClick)}
+          {showEnv && (
```

Comment:
Let's add a function here to make this more readable `this.displayEnvironment(showEnv)` or something along those lines.